File: backend/schemas/hops_schema.py
from marshmallow import Schema, fields, validates, ValidationError, EXCLUDE

class HopsSchema(Schema):
    class Meta:
        unknown = EXCLUDE

    itemUserId = fields.Int(load_only=True)
    id = fields.Int(dump_only=True)
    name = fields.Str(required=True)
    supplier = fields.Str(allow_none=True)
    alphaAcidContent = fields.Decimal(as_string=True, allow_none=True)
    betaAcidContent = fields.Decimal(as_string=True, allow_none=True)
    type = fields.Str(required=True)
    countryOfOrigin = fields.Str(allow_none=True)
    description = fields.Str(allow_none=True)
    useType = fields.Str(required=True)

    VALID_TYPES = {"Pellet", "Whole", "Cryo", "CO2 Extract"}
    VALID_USE_TYPES = {"Boil", "Dry Hop", "Aroma", "Mash", "First Wort"}

    @validates("name")
    def validate_name(self, value, *args, **kwargs):
        if not value.strip():
            raise ValidationError("Name must not be empty.")
        if len(value) > 40:
            raise ValidationError("Name must be at most 40 characters long.")

    @validates("supplier")
    def validate_supplier(self, value, *args, **kwargs):
        if value and len(value) > 40:
            raise ValidationError("Supplier name must be at most 40 characters long.")

    @validates("alphaAcidContent")
    def validate_alpha_acid_content(self, value, *args, **kwargs):
        if value is not None and (value < 0 or value > 30):
            raise ValidationError("Alpha acid content must be between 0 and 30%.")

    @validates("betaAcidContent")
    def validate_beta_acid_content(self, value, *args, **kwargs):
        if value is not None and (value < 0 or value > 30):
            raise ValidationError("Beta acid content must be between 0 and 30%.")

    @validates("type")
    def validate_type(self, value, *args, **kwargs):
        if not value.strip():
            raise ValidationError("Type is required and cannot be empty.")
        if value and value not in self.VALID_TYPES:
            raise ValidationError(f"Type must be one of: {', '.join(self.VALID_TYPES)}.")

    @validates("countryOfOrigin")
    def validate_country_of_origin(self, value, *args, **kwargs):
        if value and len(value) > 50:
            raise ValidationError("Country of origin must be at most 50 characters long.")

    @validates("useType")
    def validate_use_type(self, value, *args, **kwargs):
        if not value.strip():
            raise ValidationError("Type is required and cannot be empty.")
        if value and value not in self.VALID_USE_TYPES:
            raise ValidationError(f"Use type must be one of: {', '.join(self.VALID_USE_TYPES)}.")

File: frontend/src/Pages/Recipes/schemas/hopsSchema.js
import * as Yup from 'yup';

const validUseTypes = [
  'Boil',
  'Dry Hop',
  'Whirlpool',
  'Mash',
  'First Wort',
];

export const addHopSchema = Yup.object().shape({
  selectedItem: Yup.object()
    .required('Please select a hop.'),
  boilTime: Yup.number()
    .typeError('Boil time must be a number.')
    .min(0, 'Boil time must be 0 or more.')
    .required('Boil time is required.'),
  alphaAcid: Yup.number()
    .typeError('Alpha acid must be a number.')
    .min(0, 'Alpha acid must be 0 or more.')
    .required('Alpha acid is required.'),
  quantity: Yup.number()
    .typeError('Quantity must be a number.')
    .positive('Quantity must be greater than zero.')
    .required('Quantity is required.'),
});

export const updateHopSchema = Yup.object().shape({
  name: Yup.string()
    .trim()
    .required('Name is required')
    .min(2, 'Name must be at least 2 characters'),
  description: Yup.string()
    .nullable(),
  alphaAcidContent: Yup.number()
    .typeError('Alpha Acid must be a number.')
    .min(0, 'Alpha Acid must be 0 or more.')
    .notRequired('Alpha Acid is required.'),
  betaAcidContent: Yup.number()
    .typeError('Beta Acid must be a number.')
    .min(0, 'Beta Acid must be 0 or more.')
    .nullable(),
  boilTime: Yup.number()
    .typeError('Boil time must be a number.')
    .min(0, 'Boil time must be 0 or more.')
    .notRequired('Boil time is required.'),
  useType: Yup.string()
    .oneOf(validUseTypes, 'Invalid use type.')
    .required('Use type is required.'),
  quantity: Yup.number()
    .typeError('Quantity must be a number.')
    .positive('Quantity must be greater than zero.')
    .required('Quantity is required.'),
});
